### DISCUSSION TITLE
Virtual Machine Primary IP Management

### DISCUSSION THREAD

**Discussion by drh-jwinters** [netbox-community/netbox#8992] [Q&A]
I'm relatively new to Netbox, so I am entirely open to the possibility that the answer here is (essentially) that what I'm doing isn't in line with how Netbox was intended to be used. 

Currently, I have our IPs configured such that when we deploy a new VM, we can simply search/select the IP:

1. Create the VM
2. Create the interface
3. Click + to add the IP
4. Click the tab to Assign IP
5. Enter the first few octets of the range
6. Click on the IP
7. Click to make it the primary IP of the VM
8. Save 

Other than wishing for Netbox to just add a single default interface to the VM, this workflow is really what I was hoping for and it works well. This method makes it so our team members can create a VM and select an existing IP address without having to create the IP address from scratch.

The problem comes in when I go to decommission that VM. 

Because the IP address is listed as the Primary IP of the VM, I have to: 

1. Go to the VM detail page
2. Click on Interfaces
3. Click on the IP
4. Edit the IP
5. Un-check "Make this the primary IP"
6. Save
7. Click to edit the IP again
8. Click the x next to the Virtual Machine and Interface names
9. Save
10. Click on the VM Name under "Assignment" to get back to the main VM object
11. Click to delete the VM

If I try to both Un-check "Make this the primary IP" *AND* click x next to Virtual Machine and Interface names, it throws an error saying that it can't disconnect the IP because it is the primary. That is why I'm doing the extra step of editing the interface twice. 

If I simply delete the VM, it deletes the IP address as well (instead of just disassociating it, leaving the IP available/searchable for reassignment later on). 

I'm wondering if there is a way to do one of two things (or both):

* Configure Netbox so that it will let me make both edits at once (Un-check "Make this the primary IP" and remove the VM and Interface assignments).
* Configure Netbox to simply disassociate the IP address (rather than delete it) when I delete the VM

---

Or, if there is another way to configure the IP addresses so that they are searchable when assigning them to a VM's interface (such that we wouldn't have to re-create the IP address object to keep it searchable after deleting the VM), then that would be fine as well.

**Comment by candlerb** [comment]
> Other than wishing for Netbox to just add a single default interface to the VM, this workflow is really what I was hoping for and it works well

See issue #1492.  You can also use a Custom Script to perform the creation of VM+interface+address in a single step:
https://github.com/netbox-community/reports/blob/master/scripts/create_vm.py

> If I try to both Un-check "Make this the primary IP" AND click x next to Virtual Machine and Interface names, it throws an error saying that it can't disconnect the IP because it is the primary. That is why I'm doing the extra step of editing the interface twice.

I do think it's a reasonable request to be able to do that with a single Save operation on the IP Address, and probably counts as a bug.  Raise a bug report on Github giving the exact series of steps to reproduce.

You didn't say what version of Netbox you're running though, so if you're not on latest 3.1.x, test it there to see if the problem still exists.

> If I simply delete the VM, it deletes the IP address as well (instead of just disassociating it, leaving the IP available/searchable for reassignment later on)

That's true, and it's normally how people use Netbox: the absence of an IP address implies the ability to assign that IP address.

You could also write a custom script which:
- removes the primary IP from the VM
- unassigns the IP from the interface (possibly also updating its status or tagging it, to identify as an orphan address)
- deletes the VM
